### ISSUE TITLE
`store:ResourcesIds` getting records before it's schema is registered in `store:Tables`

### ISSUE THREAD

**Issue by hanchon** [latticexyz/mud#2061]
# Issue
Records are being set in the `ResourceIds` table before emitting the event to register that table in `Tables`. 
Without that event indexers are not able to read the `ResourcesIds` schema from the chain, forcing the dev to manually add the schema in the code.

When deploying a world, the first five events are always the same:

1. SetRecord in {tb store Tables} with key: 0x746273746f72650000000000000000005461626c657300000000000000000000 `tb store Tables`. 
- _This one creates the tbstoreTables table_

2. Splice static record in {tb store ResourceIds} with key: 0x746273746f72650000000000000000005461626c657300000000000000000000 `tb store Tables`
- _At this point the ResourceIds table schema event was not sent, so it can not be processed by just reading the events_

3. SetRecord in {tb store Tables} with key: 0x746273746f726500000000000000000053746f7265486f6f6b73000000000000 `tb store StoreHooks`
- _This one creates the tbstoreStoreHooks table_

4. Splice static record in {tb store ResourceIds} with key: 0x746273746f726500000000000000000053746f7265486f6f6b73000000000000 `tb store StoreHooks`
- _At this point the ResourceIds table schema event was not sent, so it can not be processed by just reading the events_

5. SetRecord in {tb store Tables} with key: 0x746273746f72650000000000000000005265736f757263654964730000000000 `tb store ResourceIds`

- After this event everything works fine because we got the `ResourceIds` schema

# Temp solution
For my indexer I'm just creating the `ResourcesIds` table as soon I get a `SpliceStaticRecord` on that table with its schema hardcoded in my code.

I don't know if it's posible to change the deployment script to register the `ResourcesIds` table as the second transaction, but that would allow the indexer to read all the information from the chain instead of needing to manually add the schemas to it.

# Env
Node v18.12.0
Mud running on branch main, current top commit: `392c4b88d033d2d175541b974189a3f4da49e335`

# Steps to reproduce it
Clone this repo, checkout to the commit listed on the env section
`pnpm install`
cd templates/react
`pnpm install && pnpm run dev`

With the frontend running, let the process run a couple of seconds and kill the contract one, this will stop the frontend from cleaning the transactions related to the world deployment.

![Screenshot 2023-12-12 at 13 16 29](https://github.com/latticexyz/mud/assets/5675227/e1985354-21a6-47b1-a375-82d021215f89)

**Comment by holic** [comment]
Yep, looks like we have a bit of a race condition.

`StoreCore`'s `registerTable` function calls methods on both `Tables` and `ResourceIds` tables:
https://github.com/latticexyz/mud/blob/8f41f81f77032b6168f2a4fb8de1dd9bd59e69ec/packages/store/src/StoreCore.sol#L214-L218

And we call `registerTable` for each of `Tables` and `ResourceIds`:
https://github.com/latticexyz/mud/blob/8f41f81f77032b6168f2a4fb8de1dd9bd59e69ec/packages/store/src/StoreCore.sol#L98-L100

So even if we re-order these, we'd be writing to one unregistered table.

cc @alvrs

**Comment by alvrs** [comment]
Yeah agreed I'm not sure if it's fully solvable - we could register `ResourceIds` before `StoreHooks` to reduce the amount of bootstrapping events, but it'll still require to hardcode the `Tables` table schema (since the first record we set is its schema), and the `ResourceIds` schema (since we have to register the `Tables` table resource id when registering that table). Alternatively you could only hard code the `Tables` table schema and resource ID and then ignore the first `ResourceIds` table record (setting the `Tables` table id), then the next event is registering the `ResourceIds` table (which will work because at that point the `Tables` table is initialized) and then its resource id (which will work because at that point the `ResourceIds` table will be initialized).

tl;dr: I think the best we can do is to reorder the registrations to register `ResourceIds` right after `Tables` and before `StoreHooks`

**Comment by holic** [comment]
I re-ordered them with a code comment here: https://github.com/latticexyz/mud/pull/2164

**Comment by holic** [comment]
Another approach could be doing the low level calls in order, so instead of

```solidity
Tables.register();
ResourceIds.register();
```

we do

```solidity
Tables._set(TablesTableId, Tables.getFieldLayout(), Tables.getKeySchema(), Tables.getValueSchema(), abi.encode(Tables.getKeyNames()), abi.encode(Tables.getFieldNames()));
Tables._set(ResourceIdsTableId, ResourceIds.getFieldLayout(), ResourceIds.getKeySchema(), ResourceIds.getValueSchema(), abi.encode(ResourceIds.getKeyNames()), abi.encode(ResourceIds.getFieldNames()));

ResourceIds._setExists(TablesTableId, true);
ResourceIds._setExists(ResourceIdsTableId, true);
```
